### PR TITLE
[clang][perf-training] Fix profiling with -DCLANG_BOLT=perf

### PR DESCRIPTION
**Pull request by tstellar**
This fixes the llvm-support build that generates the profile data, and wraps the whole `cmake --build` command with perf instead of wrapping each individual clang invocation.  This limits the number of profile files generated and reduces the time spent running perf2bolt.

### CODE REVIEW

**Review by aaupov — APPROVED** [review]
LG

**Comment by aaupov** [comment]
> However, I'm wondering if maybe we should disable llvm-suppot and only run hello-world with -DCLANG_BOLT=perf. 

Existing perf training is inadequate for collecting sampled profile – we simply don't get enough samples, and no-LBR mode further drops the performance. If we wanted to pursue perf sampling further, we'd need to extend perf training with either building LLVM subtargets or llvm-test-suite.

> The bolt optimizations with perf only give about a 3% performance increase (although maybe with hw counters this would be better) 

Yes, no-LBR mode has very limited benefit due to missing edge counts.

> and it takes a very long time to convert all the perf profiles to the fdata format.

As discussed on Discord, we may be able to reduce the time by dropping `profile-format=yaml` which is not required in this case. By the way, although I understand the overhead adds up, we shouldn't be converting many profiles: I guess just two - one for cxx/helloworld and another for llvm-support.

**Comment by tstellar** [comment]
@aaupov When we build llvm-support there is one perf.data file generate for each cpp file compiled, so we end up without about 150 files.  Is there some way to merge those together before running perf2bolt?

**Comment by aaupov** [comment]
> @aaupov When we build llvm-support there is one perf.data file generate for each cpp file compiled, so we end up without about 150 files.  Is there some way to merge those together before running perf2bolt?

I see. The best way would be to run perf once so that all clang invocations are under it. If llvm-support is configured as cmake rule for llvm external project, perf wrapper could be set as CMAKE_COMPILER_LAUNCHER.

**Comment by tstellar** [comment]
> > @aaupov When we build llvm-support there is one perf.data file generate for each cpp file compiled, so we end up without about 150 files.  Is there some way to merge those together before running perf2bolt?
> 
> I see. The best way would be to run perf once so that all clang invocations are under it. If llvm-support is configured as cmake rule for llvm external project, perf wrapper could be set as CMAKE_COMPILER_LAUNCHER.

Even with using CMAKE_CXX_COMPILER_LAUNCHER, we still get one invocation of perf for each invocation of clang.  I just tested out `%perf_wrapper %cmake --build %t -v --target LLVMSupport` and that seems to work.

**Review by aaupov — APPROVED** [review]

File: clang/utils/perf-training/bolt.lit.cfg [hunk 1]
```diff
@@ -20,9 +21,13 @@ elif clang_bolt_mode == "perf":
 else:
     assert 0, "Unsupported CLANG_BOLT_MODE variable"
 
-config.clang = perf_wrapper + os.path.realpath(
+clang_nowrapper = os.path.realpath(
     lit.util.which(clang_binary, config.clang_tools_dir)
 ).replace("\\", "/")
+config.clang = perf_wrapper + clang_nowrapper
```
Comment (aaupov) [inline clang/utils/perf-training/bolt.lit.cfg:27]: Just for my understanding: we're keeping perf_wrapper set if CLANG_BOLT=perf. However we pass perf_wrapper in front of cmake command in `clang/utils/perf-training/llvm-support/build.test` – do we end up with nested perf invocations, or do we strip perf_wrapper from clang in this case?

File: clang/utils/perf-training/bolt.lit.cfg [hunk 2]
```diff
@@ -20,9 +21,13 @@ elif clang_bolt_mode == "perf":
...
+config.cmake_compiler_args = "-DCMAKE_C_COMPILER='{0}' -DCMAKE_CXX_COMPILER='{0};--driver-mode=g++'".format(
+    re.sub(r"\s+", ";", clang_nowrapper)
+)
```
Comment (tstellar) [inline clang/utils/perf-training/bolt.lit.cfg:30]: @aaupov Here we are using clang_nowrapper as the compiler, so we don't end up with nested perf calls.

**Pull request merged**